src/class/audio/audio_device.c

```diff
@@ -519,199 +811,189 @@ static bool audiod_tx_done_cb(uint8_t rhport, audiod_interface_t* audio, uint16_
-
-  return true;
-}
-#endif // CFG_TUD_AUDIO_TX_FIFO_COUNT > 1 || (CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX != CFG_TUD_AUDIO_TX_ITEMSIZE)
+  // Check if there is enough
+  if (nBytesPerFFToSend == 0)    return 0;
 
-#endif //CFG_TUD_AUDIO_TX_FIFO_SIZE
+  // Limit to maximum sample number - THIS IS A POSSIBLE ERROR SOURCE IF TOO MANY SAMPLE WOULD NEED TO BE SENT BUT CAN NOT!
+  nBytesPerFFToSend = tu_min16(nBytesPerFFToSend, capPerFF);
 
-// This function is called once a transmit of an feedback packet was successfully completed. Here, we get the next feedback value to be sent
+  // Round to full number of samples (flooring)
+  nBytesPerFFToSend = (nBytesPerFFToSend / nBytesToCopy) * nBytesToCopy;
 
-#if CFG_TUD_AUDIO_EPSIZE_OUT && CFG_TUD_AUDIO_ENABLE_FEEDBACK_EP
-static bool audio_fb_send(uint8_t rhport, audiod_interface_t *audio)
-{
-  uint8_t fb[4];
+  // Encode
+  void * src;
+  uint8_t * dst;
+  uint8_t * src_end;
   uint16_t len;
 
-  if (audio->fb_val == 0)
-  {
-    len = 0;
-    return true;
-  }
-  else
+  for (cnt_ff = 0; cnt_ff < n_ff_used; cnt_ff++)
   {
-    len = 4;
-    // Here we need to return the feedback value
-    if (rhport == 0)
-    {
-      // For FS format is 10.14
-      fb[0] = (audio->fb_val >> 2) & 0xFF;
-      fb[1] = (audio->fb_val >> 10) & 0xFF;
-      fb[2] = (audio->fb_val >> 18) & 0xFF;
-      // 4th byte is needed to work correctly with MS Windows
-      fb[3] = 0;
-    }
-    else
-    {
-      // For HS format is 16.16
-      fb[0] = (audio->fb_val >> 0) & 0xFF;
-      fb[1] = (audio->fb_val >> 8) & 0xFF;
-      fb[2] = (audio->fb_val >> 16) & 0xFF;
-      fb[3] = (audio->fb_val >> 24) & 0xFF;
-    }
-    return usbd_edpt_xfer(rhport, audio->ep_fb, fb, len);
-  }
-
-}
+    dst = &audio->lin_buf_in[cnt_ff*audio->n_channels_per_ff_tx*audio->n_bytes_per_sampe_tx];
 
-//static uint16_t audio_fb_done_cb(uint8_t rhport, audiod_interface_t* audio)
-//{
-//  (void) rhport;
-//  (void) audio;
-//
-//  if (tud_audio_fb_done_cb) TU_VERIFY(tud_audio_fb_done_cb(rhport));
-//  return 0;
-//}
+    len = tu_fifo_get_linear_read_info(&audio->tx_supp_ff[cnt_ff], 0, &src, nBytesPerFFToSend);
+    tu_fifo_advance_read_pointer(&audio->tx_supp_ff[cnt_ff], len);
 
-#endif
+    src_end = src + len;
 
-// This function is called once a transmit of an interrupt control packet was successfully completed. Here, we get the remaining bytes to send
+    dst = audiod_interleaved_copy_bytes_fast_encode(nBytesToCopy, src, src_end, dst, n_ff_used);
 
-#if CFG_TUD_AUDIO_INT_CTR_EPSIZE_IN
-static bool audio_int_ctr_done_cb(uint8_t rhport, audiod_interface_t* audio, uint16_t * n_bytes_copied)
-{
-  // We write directly into the EP's buffer - abort if previous transfer not complete
-  TU_VERIFY(!usbd_edpt_busy(rhport, audio->ep_int_ctr));
+    // Handle wrapped part of FIFO
+    if (len < nBytesPerFFToSend)
+    {
+      len = tu_fifo_get_linear_read_info(&audio->tx_supp_ff[cnt_ff], 0, &src, nBytesPerFFToSend - len);
```

Comment:
This is an optimized code which conducts interleaved sample copies! I would strongly advise against doing this with write_n/read_n because it is dramatically inefficient! The way it is done here avoids the unnecessary overhead of updating the read and write pointers for every sample and furthermore, it does not use memcpy for the copy of 1,2,3, or 4 bytes (which by definition are all possible sample sizes in UAC2 type I format).

I understand your desire to put such code into fifo.c but this is a very special purpose here and only required for the audio encoding/deconding. I am not sure if it is good to shift this into fifo.c! In any case please do not put that away! The CPU load would increase unnecessarily!!

Reply:
Ah ok, got it, thanks for detail explanation